### ISSUE TITLE
Clippy --fix panics with index error

### ISSUE THREAD

**Issue by tvallotton** [rust-lang/rust-clippy#8734]
### Summary

`cargo clippy --fix` causes an index error.

### Reproducer

I pushed the code that causes the error to [this repo](https://github.com/tvallotton/chess), so you can reproduce it. The error was triggerd by `cargo clippy --fix`. Note that this does not seem to panic in stable. 

```rust
thread 'main' panicked at 'slice index starts at 1 but ends at 0', library/core/src/slice/index.rs:91:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: could not compile `chess`
warning: build failed, waiting for other jobs to finish...
thread 'main' panicked at 'slice index starts at 1 but ends at 0', library/core/src/slice/index.rs:91:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: could not compile `chess`
```

### Version

```text
rustc 1.61.0-nightly (76d770ac2 2022-04-02)
binary: rustc
commit-hash: 76d770ac21d9521db6a92a48c7b3d5b2cc535941
commit-date: 2022-04-02
host: aarch64-apple-darwin
release: 1.61.0-nightly
LLVM version: 14.0.0
```

**Comment by tvallotton** [comment]
it looks like its related to #8718.

**Comment by xFrednet** [comment]
Hey, thank you for the report. This looks like the linked issue. Could you provide the Stacktrace, by setting `RUST_BACKTRACE=1`? :upside_down_face:

**Comment by tvallotton** [comment]
I get. You can also try to replicate it yourself on [this repo](https://github.com/tvallotton/chess/tree/clippy-fix-panic)
```rust
thread 'main' panicked at 'slice index starts at 1 but ends at 0', library/core/src/slice/index.rs:91:5
stack backtrace:
   0: _rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::slice::index::slice_index_order_fail_rt
   3: core::ops::function::FnOnce::call_once
   4: core::intrinsics::const_eval_select
   5: core::slice::index::slice_index_order_fail
   6: rustfix::parse_snippet
   7: rustfix::collect_span
   8: <alloc::vec::Vec<rustfix::Replacement> as alloc::vec::spec_from_iter::SpecFromIter<rustfix::Replacement, core::iter::adapters::filter_map::FilterMap<core::iter::adapters::filter::Filter<core::slice::iter::Iter<rustfix::diagnostics::DiagnosticSpan>, rustfix::collect_suggestions<std::collections::hash::map::RandomState>::{closure#1}::{closure#0}>, rustfix::collect_span>>>::from_iter
   9: <&mut rustfix::collect_suggestions<std::collections::hash::map::RandomState>::{closure#1} as core::ops::function::FnMut<(&rustfix::diagnostics::Diagnostic,)>>::call_mut
  10: <alloc::vec::Vec<rustfix::Solution> as alloc::vec::spec_from_iter::SpecFromIter<rustfix::Solution, core::iter::adapters::filter_map::FilterMap<core::slice::iter::Iter<rustfix::diagnostics::Diagnostic>, rustfix::collect_suggestions<std::collections::hash::map::RandomState>::{closure#1}>>>::from_iter
  11: rustfix::collect_suggestions::<std::collections::hash::map::RandomState>
  12: cargo::ops::fix::fix_maybe_exec_rustc
  13: cargo::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
error: could not compile `chess`
warning: build failed, waiting for other jobs to finish...
thread 'main' panicked at 'slice index starts at 1 but ends at 0', library/core/src/slice/index.rs:91:5
stack backtrace:
   0: _rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::slice::index::slice_index_order_fail_rt
   3: core::ops::function::FnOnce::call_once
   4: core::intrinsics::const_eval_select
   5: core::slice::index::slice_index_order_fail
   6: rustfix::parse_snippet
   7: rustfix::collect_span
   8: <alloc::vec::Vec<rustfix::Replacement> as alloc::vec::spec_from_iter::SpecFromIter<rustfix::Replacement, core::iter::adapters::filter_map::FilterMap<core::iter::adapters::filter::Filter<core::slice::iter::Iter<rustfix::diagnostics::DiagnosticSpan>, rustfix::collect_suggestions<std::collections::hash::map::RandomState>::{closure#1}::{closure#0}>, rustfix::collect_span>>>::from_iter
   9: <&mut rustfix::collect_suggestions<std::collections::hash::map::RandomState>::{closure#1} as core::ops::function::FnMut<(&rustfix::diagnostics::Diagnostic,)>>::call_mut
  10: <alloc::vec::Vec<rustfix::Solution> as alloc::vec::spec_from_iter::SpecFromIter<rustfix::Solution, core::iter::adapters::filter_map::FilterMap<core::slice::iter::Iter<rustfix::diagnostics::Diagnostic>, rustfix::collect_suggestions<std::collections::hash::map::RandomState>::{closure#1}>>>::from_iter
  11: rustfix::collect_suggestions::<std::collections::hash::map::RandomState>
  12: cargo::ops::fix::fix_maybe_exec_rustc
  13: cargo::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
error: could not compile `chess`
```

**Comment by Serial-ATA** [comment]
Looks like it's coming from the `map_flatten` lint here: https://github.com/tvallotton/chess/blob/98b01208e95173388cf4f781cc83bb414bb3bf55/src/table.rs#L73

It doesn't like the long multiline suggestions.

For example, this works:

```rust
[0u8, 1, 2, 3]
    .into_iter()
    .map(|n| match n {
        1 => [n.saturating_add(1)],
        n => [n],
    })
    .flatten();
```

While this causes the panic:

```rust
[0u8, 1, 2, 3]
    .into_iter()
    .map(|n| match n {
        1 => [n
            .saturating_add(1)
            .saturating_add(1)
            .saturating_add(1)
            .saturating_add(1)
            .saturating_add(1)
            .saturating_add(1)
            .saturating_add(1)
            .saturating_add(1)],
        n => [n],
    })
    .flatten();
```

Works just fine in stable, so it's likely #8520

**Comment by Serial-ATA** [comment]
This is actually a Cargo issue. https://github.com/rust-lang/rustfix/commit/635eef76315c389017b1d41f443db50c89a8a2ba didn't make it into rustfix 0.6.0, which Cargo depends on. Just have to wait for a new release.

**Comment by droogmic** [comment]
I ran into this as well, I see we are waiting for an answer on https://github.com/rust-lang/rustfix/issues/211.

**Comment by hellow554** [comment]
The culprit should be solved by updating rustfix to 0.6.1, but @Serial-ATA discovered a very interesting case:

```rust
fn issue8734() {
    let _ = [0u8, 1, 2, 3]
        .into_iter()
        .map(|n| match n {
            1 => [n
                .saturating_add(1)
                .saturating_add(1)
                .saturating_add(1)
                .saturating_add(1)
                .saturating_add(1)
                .saturating_add(1)
                .saturating_add(1)
                .saturating_add(1)],
            n => [n],
        })
        .flatten();
}
```

gets autocorrect to

```rust
fn issue8734() {
    let _ = [0u8, 1, 2, 3]
        .into_iter()
        .flat_map(|n| match n {
            1 => [n
                .saturating_add(1)
            1 => [n
                .saturating_add(1)
                .saturating_add(1)
                .saturating_add(1)
                .saturating_add(1)
                .saturating_add(1)
                .saturating_add(1)
                .saturating_add(1)
                .saturating_add(1)],
            n => [n],
        });
}
```

which is weird and incorrect. 

Let's investigate this further
